src/main/java/ca/uhn/fhir/jpa/starter/interceptor/CustomAuthorizationInterceptor.java

```diff
@@ -0,0 +1,179 @@
+		return isOAuthEnabled() || isApiKeyEnabled();
+	}
+
+	private boolean isUsingOAuth(RequestDetails theRequest) {
+		return isOAuthEnabled() && OAuth2Helper.hasToken(theRequest);
+	}
+
+	private boolean isUsingApiKey(RequestDetails theRequest) {
+		return isApiKeyEnabled() && ApiKeyHelper.hasApiKey(theRequest);
+	}
+
+	private List<IAuthRule> authorizedRule() {
+		return new RuleBuilder()
+			.allowAll()
+			.build();
+	}
+
+	private List<IAuthRule> unauthorizedRule() {
+		// By default, deny everything except the metadata request
+		return new RuleBuilder()
+			.allow().metadata().andThen()
+			.denyAll()
+			.build();
+	}
+
+	private List<IAuthRule> authorizeOAuth(RequestDetails theRequest) throws AuthenticationException {
+		logger.info("Authorizing via OAuth2");
+		String token = OAuth2Helper.getToken(theRequest);
+		try {
+			DecodedJWT jwt = JWT.decode(token);
+			String jwksUrl = config.getOauth().getJwks_url();
+			OAuth2Helper.verify(jwt, jwksUrl);
+
+			List<String> clientRoles = OAuth2Helper.getClientRoles(jwt, getOAuthClientId());
+			if (clientRoles.isEmpty()) {
+				logger.warn("Authorization failure - token doesn't have any client roles");
+				return unauthorizedRule();
+			}
+
+			// The only difference between the admin role and the user role is that the admin role
+			// allows DELETE requests. It still needs to enforce a patient claim, if one exists.
+			if (theRequest.getRequestType().equals(RequestTypeEnum.DELETE)
+					&& !clientRoles.contains(getOAuthAdminRole())) {
+				logger.warn("Authorization failure - token doesn't have the admin role required for delete");
+				return unauthorizedRule();
+			}
+
+			if (clientRoles.contains(getOAuthAdminRole()) || clientRoles.contains(getOAuthUserRole())) {
+				String patientId = OAuth2Helper.getClaimAsString(jwt, "patient");
+				if (Strings.isNullOrEmpty(patientId)) {
+					logger.debug("No patient claim specified in authorization token");
+					return authorizedRule();
+				} else {
+					logger.debug("Patient claim specified in in authorization token; will use patient compartment rules");
+					return authorizedInPatientCompartmentRule(theRequest, patientId);
+				}
+			}
+
+			logger.warn("Authorization failure - token doesn't have the required client roles");
+			return unauthorizedRule();
+		} catch (RuntimeException|GeneralSecurityException e) {
+			logger.warn("Authentication failure - unable to decode or verify token: {}", e.getMessage());
+			throw new AuthenticationException("Invalid authorization header", e);
+		}
+	}
+
+	private List<IAuthRule> authorizedInPatientCompartmentRule(RequestDetails theRequestDetails, String patientId) {
+		if (OAuth2Helper.canBeInPatientCompartment(theRequestDetails.getResourceName())) {
+			IdType patientIdType = new IdType("Patient", patientId);
+			return new RuleBuilder()
+				.allow().read().allResources().inCompartment("Patient", patientIdType).andThen()
+				.allow().patch().allRequests().andThen()
```

Comment:
shouldn't we also enable inCompartment for patch request ? Currently we are not using patch request but there is possibility where data can be updated using patch request.

Reply:
It's not possible to allow patching in the patient compartment only.

https://hapifhir.io/hapi-fhir/docs/security/authorization_interceptor.html